# Assignment 2: Containers

- Please name your files:
    * ASSIGNMENT_2_FIRSTNAME_LASTNAME.ipynb 
    * assignment2_utils.py
- Please store the two files in a folder called ASSIGNMENT_2_FIRSTNAME_LASTNAME
- Please zip your folder and please follow the following naming convention for the zip file: ASSIGNMENT_2_FIRSTNAME_LASTNAME.zip
- Please submit your assignment on Canvas: Assignment 2
- If you have **questions** about this topic, please contact us **(cltl.python.course@gmail.com)**. Questions and answers will be collected on Piazza, so please check if your question has already been answered first.

In this block, we covered the following chapters:
- Chapter 05 - Core concepts of containers
- Chapter 06 - Lists
- Chapter 07 - Sets
- Chapter 08 - Comparison of lists and sets
- Chapter 09 - Looping over containers.
- Chapter 10 - Dictionaries
- Chapter 11 - Functions and scope

In this assignment, you will be asked to show what you have learned from the topics above! 

**Finding solutions online**

Very often, you can find good solutions online. We encourage you to use online resources when you get stuck. However, please always try to understand the code you find and indicate that it is not your own. Use the following format to mark code written by someone else:

###Taken from [link] [date]

[code]

\###

Please use a similar format to indicate that you have worked with a classmate (e.g. mention the name instead of the link). 

*Please stick to this strategy for all course assignments.*

## Exercise 1: Beersong
*99 Bottles of Beer* is a traditional song in the United States and Canada. Write a python program that generates the lyrics to the song.

The song's simple lyrics are as follows:

    99 bottles of beer on the wall,
    99 bottles of beer.
    Take one down, pass it around,
    98 bottles of beer on the wall.

The same verse is repeated, each time with one fewer bottle. The song
is completed when the singer or singers reach zero. After the last bottle
is taken down and passed around, there is a special verse:

    No more bottles of beer on the wall,
    no more bottles of beer.
    Go to the store and buy some more,
    99 bottles of beer on the wall.

Notes:

* Leave a blank line between verses.
* Make sure that you print the singular form of "bottles" when the counter is at one.

Hint:

* While debugging the program, start from a small number, and
change it to 100 when you are done (as shown below).
* Use variables to prevent code repetition

You can use the following code snippet as a start:

In [25]:
starting_num_bottles = 99 # the total number of bottles to start with

for bottles in range(starting_num_bottles, -1, -1): # end range function at -1 to include 0 bottles
    remaining_num_bottles = bottles - 1
    # initialize text used for the majority of the verses
    text_bottles = "bottles"
    text_remaining_bottles = "bottles"
    key_sentence = "Take one down, pass it around," # the only sentence that changes in the last verse

    # change text accordingly to the current number of bottles
    if bottles == 2:
        text_remaining_bottles = "bottle"
    elif bottles == 1:
        text_bottles = "bottle"
        remaining_num_bottles = "no more"
    elif bottles == 0: # change text accordingly for the special final verse
        bottles = "No more"
        key_sentence = "Go to the store and buy some more," # change key sentence
        remaining_num_bottles = starting_num_bottles # change to inital number of bottles
    
    print(f"{bottles} {text_bottles} of beer on the wall,")
    print(f"{'no more' if bottles == 'No more' else bottles} {text_bottles} of beer.") # get rid of capital letter in final verse
    print(f"{key_sentence}")
    print(f"{remaining_num_bottles} {text_remaining_bottles} of beer on the wall.")
    print()

99 bottles of beer on the wall,
99 bottles of beer.
Take one down, pass it around,
98 bottles of beer on the wall.

98 bottles of beer on the wall,
98 bottles of beer.
Take one down, pass it around,
97 bottles of beer on the wall.

97 bottles of beer on the wall,
97 bottles of beer.
Take one down, pass it around,
96 bottles of beer on the wall.

96 bottles of beer on the wall,
96 bottles of beer.
Take one down, pass it around,
95 bottles of beer on the wall.

95 bottles of beer on the wall,
95 bottles of beer.
Take one down, pass it around,
94 bottles of beer on the wall.

94 bottles of beer on the wall,
94 bottles of beer.
Take one down, pass it around,
93 bottles of beer on the wall.

93 bottles of beer on the wall,
93 bottles of beer.
Take one down, pass it around,
92 bottles of beer on the wall.

92 bottles of beer on the wall,
92 bottles of beer.
Take one down, pass it around,
91 bottles of beer on the wall.

91 bottles of beer on the wall,
91 bottles of beer.
Take one down, pass 

## Exercise 2: list methods

In this exercise, we will focus on the following **list methods**:

a) append <br>
b) count <br>
c) index <br>
d) insert <br>
e) pop <br>

For each of the aforementioned list methods:

* explain the positional parameters
* explain the keyword parameters
* you can exclude *self* from your explanation
* explain what the goal of the method is and what data type it returns, e.g., string, list, set, etc.
* give a working example. Provide also an example by providing a value for keyword parameters (assuming the method has one or more keyword parameters).

### Please fill in your answers here: 


### a) append

In [83]:
help(list.append)

Help on method_descriptor:

append(self, object, /)
    Append object to the end of the list.



#### append: positional parameters

1) object: the list method append has one positional parameter, namely object. The value of the argument assigned to this parameter is the object you want to add to the end of your list.

#### append: keyword parameters

The list method append does not have any keyword parameters.

#### append: goal of the method

The goal of this method is, as previously mentioned, to add (or append) an object to the end of a list. This method modifies and already existing list and does not return anything. However, lists can contain any objects, so objects of all data types can be added.

#### append: a working example

In [84]:
list_of_brand_names = ["Nike", "Apple", "Coca Cola", "McDonalds"]

# if we print this we get the following
print(list_of_brand_names)

['Nike', 'Apple', 'Coca Cola', 'McDonalds']


In [85]:
# now we want to append the brandname "Samsung" to the end of the list and therefore we use append
list_of_brand_names.append("Samsung")

# and we print the new list, observing that "Samsung" is added to the end
print(list_of_brand_names)

['Nike', 'Apple', 'Coca Cola', 'McDonalds', 'Samsung']


#### b) count

In [69]:
help(list.count)

Help on method_descriptor:

count(self, value, /)
    Return number of occurrences of value.



#### count: positional parameters

1) value: the list method count has one positional parameter, namely value. The element that you wish to count in a given list is the value of the argument assigned to this parameter.

#### count: keyword parameters

The list method count does not have any keyword parameters.

#### count: goal of the method

The goal of this method is to count the number of times a given element occurs in a list. The method returns this number as an integer.

#### count: a working example

In [88]:
list_of_student_names = ["Ahmed", "Bob", "Ahmed", "Sarah", "Kim", "Sascha"]

# So, we have the following list
print(list_of_student_names)

['Ahmed', 'Bob', 'Ahmed', 'Sarah', 'Kim', 'Sascha']


In [90]:
# Assume we want to know how many students are called Ahmed, we can use the count method
count_of_ahmed = list_of_student_names.count("Ahmed")
print(count_of_ahmed)

# We see that the count is 2

2


#### c) index

In [70]:
help(list.index)

Help on method_descriptor:

index(self, value, start=0, stop=9223372036854775807, /)
    Return first index of value.
    
    Raises ValueError if the value is not present.



#### index: positional parameters

1) value: the value of the argument assigned to this parameter is the element for which you want to know the index of its first occurrence.
2) start: the index at which you wish to begin looking for the element is the value of the argument that was assigned to this parameter. The default value of this is 0.
3) stop: this is the index at which you wish to stop searching. The index you assign to this parameter is not included in the search, so you search up to this point. The default value is 9223372036854775807, so a very high number.

#### index: keyword parameters

The list method index does not have any keyword parameters.

#### index: goal of the method

The goal of this method is to find the index of the first occurence of an element you are looking for in a list. The method return the index of this element, which is an integer number. 

#### index: a working example

In [95]:
list_of_football_clubs = ["Manchester United", "Ajax", "FC Barcelona", "Bayern Munchen", "Real Madrid"]

# we get the following list
print(list_of_football_clubs)

['Manchester United', 'Ajax', 'FC Barcelona', 'Bayern Munchen', 'Real Madrid']


In [96]:
# now we want to know the index at which Ajax occurs
index_ajax = list_of_football_clubs.index("Ajax")
print(index_ajax)

# the index at which Ajax occurs is 1

1


#### d) insert

In [110]:
help(list.insert)

Help on method_descriptor:

insert(self, index, object, /)
    Insert object before index.



#### insert: positional parameters

1) index: The value of the argument assigned to this parameter is the index at which you like to insert the object before.
2) object: This is an element you would like to insert to your list. 

#### insert: keyword parameters

The list method insert does not have any keyword parameters.

#### insert: goal of the method

The goal of this method is to insert an element before a given index in a list. This method modifies an already existing list and does not return anything. 

#### insert: a working example

In [111]:
list_of_city_names = ["Berlin", "Madrid", "Rome"]

# this gives us the following list
print(list_of_city_names)

['Berlin', 'Madrid', 'Rome']


In [112]:
# assume we want to insert Amsterdam but also keep the list in alphabetical order, we can do the following:
list_of_city_names.insert(0, "Amsterdam")

# Amsterdam is added to the list
print(list_of_city_names)

['Amsterdam', 'Berlin', 'Madrid', 'Rome']


#### e) pop

In [113]:
help(list.pop)

Help on method_descriptor:

pop(self, index=-1, /)
    Remove and return item at index (default last).
    
    Raises IndexError if list is empty or index is out of range.



#### pop: positional parameters

1) index: the value of this argument assigned to the parameter is the index of the element you would like to remove from your list. The default value is -1, so the last element of the list.

#### pop: keyword parameters

The list method pop does not have any keyword parameters.

#### pop: goal of the method

The goal of this method is to remove an element of a list at a given index and return this element. The data type of the returned element could be any, it depends on the data type of the element. 

#### pop: a working example

In [122]:
list_of_number = [2.78, 3.1415, 0, 10, 123]

# this will give us the following list
print(list_of_number)

[2.78, 3.1415, 0, 10, 123]


In [123]:
# assume we want to remove pi (on index 1) from this list
pi = list_of_number.pop(1)

# printing the new list, we see that pi is removed and assigned to the variable pi
print(list_of_number)
print(pi)

[2.78, 0, 10, 123]
3.1415


## Exercise 3: set methods

In this exercise, we will focus on the following **set methods**:

a) update <br>
b) pop <br>
c) remove <br>
d) clear <br>

For each of the aforementioned set methods:

* explain the positional parameters
* explain the keyword parameters
* you can exclude *self* from your explanation
* explain what the goal of the method is and what data type it returns, e.g., string, list, set, etc.
* give a working example. Provide also an example by providing a value for keyword parameters (assuming the method has one or more keyword parameters).

### Please fill in your answers here: 


#### a) update

In [106]:
help(set.update)

Help on method_descriptor:

update(...)
    Update a set with the union of itself and others.



#### b) pop

In [107]:
help(set.pop)

Help on method_descriptor:

pop(...)
    Remove and return an arbitrary set element.
    Raises KeyError if the set is empty.



#### c) remove

In [108]:
help(set.remove)

Help on method_descriptor:

remove(...)
    Remove an element from a set; it must be a member.
    
    If the element is not a member, raise a KeyError.



#### d) clear

In [109]:
help(set.clear)

Help on method_descriptor:

clear(...)
    Remove all elements from this set.



## Exercise 4: Analyzing vocabulary using sets
Please consider the following two texts:

These stories were copied from [here](http://www.english-for-students.com/).

In [26]:
a_story = """In a far away kingdom, there was a river. This river was home to many golden swans. The swans spent most of their time on the banks of the river. Every six months, the swans would leave a golden feather as a fee for using the lake. The soldiers of the kingdom would collect the feathers and deposit them in the royal treasury. 
One day, a homeless bird saw the river. "The water in this river seems so cool and soothing. I will make my home here," thought the bird. 
As soon as the bird settled down near the river, the golden swans noticed her. They came shouting. "This river belongs to us. We pay a golden feather to the King to use this river. You can not live here." 
"I am homeless, brothers. I too will pay the rent. Please give me shelter," the bird pleaded. "How will you pay the rent? You do not have golden feathers," said the swans laughing. They further added, "Stop dreaming and leave once." The humble bird pleaded many times. But the arrogant swans drove the bird away. 
"I will teach them a lesson!" decided the humiliated bird. 
She went to the King and said, "O King! The swans in your river are impolite and unkind. I begged for shelter but they said that they had purchased the river with golden feathers." 
The King was angry with the arrogant swans for having insulted the homeless bird. He ordered his soldiers to bring the arrogant swans to his court. In no time, all the golden swans were brought to the King’s court. 
"Do you think the royal treasury depends upon your golden feathers? You can not decide who lives by the river. Leave the river at once or you all will be beheaded!" shouted the King. 
The swans shivered with fear on hearing the King. They flew away never to return. The bird built her home near the river and lived there happily forever. The bird gave shelter to all other birds in the river. """
print(a_story)

In a far away kingdom, there was a river. This river was home to many golden swans. The swans spent most of their time on the banks of the river. Every six months, the swans would leave a golden feather as a fee for using the lake. The soldiers of the kingdom would collect the feathers and deposit them in the royal treasury. 
One day, a homeless bird saw the river. "The water in this river seems so cool and soothing. I will make my home here," thought the bird. 
As soon as the bird settled down near the river, the golden swans noticed her. They came shouting. "This river belongs to us. We pay a golden feather to the King to use this river. You can not live here." 
"I am homeless, brothers. I too will pay the rent. Please give me shelter," the bird pleaded. "How will you pay the rent? You do not have golden feathers," said the swans laughing. They further added, "Stop dreaming and leave once." The humble bird pleaded many times. But the arrogant swans drove the bird away. 
"I will teach

In [27]:
another_story = """Long time ago, there lived a King. He was lazy and liked all the comforts of life. He never carried out his duties as a King. "Our King does not take care of our needs. He also ignores the affairs of his kingdom." The people complained. 
One day, the King went into the forest to hunt. After having wandered for quite sometime, he became thirsty. To his relief, he spotted a lake. As he was drinking water, he suddenly saw a golden swan come out of the lake and perch on a stone. "Oh! A golden swan. I must capture it," thought the King. 
But as soon as he held his bow up, the swan disappeared. And the King heard a voice, "I am the Golden Swan. If you want to capture me, you must come to heaven." 
Surprised, the King said, "Please show me the way to heaven." Do good deeds, serve your people and the messenger from heaven would come to fetch you to heaven," replied the voice. 
The selfish King, eager to capture the Swan, tried doing some good deeds in his Kingdom. "Now, I suppose a messenger will come to take me to heaven," he thought. But, no messenger came. 
The King then disguised himself and went out into the street. There he tried helping an old man. But the old man became angry and said, "You need not try to help. I am in this miserable state because of out selfish King. He has done nothing for his people." 
Suddenly, the King heard the golden swan’s voice, "Do good deeds and you will come to heaven." It dawned on the King that by doing selfish acts, he will not go to heaven. 
He realized that his people needed him and carrying out his duties was the only way to heaven. After that day he became a responsible King. 
"""

### Exercise 4a: preprocessing text
Before analyzing the two texts, we are first going to preprocess them. 
Please use a particular string method multiple times to replace the following characters by empty strings in both **a_story** and **another_story**:
* newlines: '\n'
* commas: ','
* dots: '.'
* quotes: '"'


Please assign the processed texts to the variables **cleaned_story** and **cleaned_another_story**.

In [28]:
# preprocessing a_story
cleaned_story = a_story.replace('\n', '')
cleaned_story = cleaned_story.replace(',', '')
cleaned_story = cleaned_story.replace('.', '')
cleaned_story = cleaned_story.replace('"', '')

# preprocessing another_story
cleaned_another_story = another_story.replace('\n', '')
cleaned_another_story = cleaned_another_story.replace(',', '')
cleaned_another_story = cleaned_another_story.replace('.', '')
cleaned_another_story = cleaned_another_story.replace('"', '')



#### Exercise 4b: from text to a list
For each text (**cleaned_story** and **cleaned_another_story**), please use a string method to convert **cleaned_story** and **cleaned_another_story** into lists by splitting using spaces. Please call the lists **list_cleaned_story** and **list_cleaned_another_story**.

In [29]:
# note: assumption here is made that with "splitting using spaces" it is meant all white spaces

# convert cleaned_story into list
list_cleaned_story = cleaned_story.split()

# convert cleaned_another_story into list
list_cleaned_another_story = cleaned_another_story.split()

#### Exercise 4c: from a list to a vocabulary (a set)
Please create a set for the words in each text by adding each word to a set. In the end, you should have two variables **vocab_a_story** and **vocab_another_story**, each containing the unique words in each story. Please use the output of Exercise 4b as the input for this exercise.

In [30]:
vocab_a_story = set()
for word in list_cleaned_story:
    vocab_a_story.add(word)

do the same for the other text

In [31]:
vocab_another_story = set()
for word in list_cleaned_another_story:
    vocab_another_story.add(word)

### Exercise 4d: analyzing vocabularies
Please analyze the vocabularies by using set methods to determine:
* which words occur in both texts
* which words only occur in **a_story**
* which words only occur in **another_story**

In [32]:
occur_in_both_text = vocab_a_story.intersection(vocab_another_story)
print(occur_in_both_text)
print(len(occur_in_both_text))
print()

occur_only_a_story = vocab_a_story.difference(vocab_another_story)
print(occur_only_a_story)
print(len(occur_only_a_story))
print()

occur_only_another_story = vocab_another_story.difference(vocab_a_story)
print(occur_only_another_story)
print(len(occur_only_another_story))

{'But', 'golden', 'all', 'soon', 'went', 'lived', 'will', 'As', 'water', 'his', 'said', 'the', 'thought', 'me', 'you', 'no', 'time', 'lake', 'to', 'not', 'King', 'this', 'He', 'by', 'I', 'One', 'in', 'having', 'angry', 'was', 'would', 'came', 'never', 'on', 'as', 'and', 'of', 'am', 'Do', 'there', 'day', 'that', 'a', 'your', 'Please', 'for', 'The', 'kingdom', 'saw', 'You'}
50

{'They', 'bird', 'down', 'forever', 'lives', 'think', 'but', 'seems', 'us', 'shelter', 'flew', 'King!', 'begged', 'be', 'they', 'give', 'How', 'other', 'have', 'pleaded', 'fear', 'too', 'humble', 'impolite', 'feather', 'at', 'with', 'who', 'brothers', 'Leave', 'bring', 'O', 'humiliated', 'are', 'shouting', 'further', 'court', 'feathers', 'here', 'treasury', 'their', 'fee', 'them', 'my', 'river', 'purchased', 'In', 'use', 'or', 'gave', 'unkind', 'This', 'settled', 'months', 'leave', 'her', 'Every', 'do', 'times', 'home', 'can', 'six', 'laughing', 'swans', 'Stop', 'ordered', 'decided', 'added', 'happily', 'banks', '

## Exercise 5: counting
Below you find a list called **words**, which is a list of strings. 

a) Please create a dictionary in which the **key** is the word, and the **value** is the frequency of the word. 


Exclude all words which meet at least one of the following requirements:

 * end with the letter e
 * start with the letter t
 * start with the letter c and end with the letter w (both conditions must be met)
 * have six or more letters

You are not allowed to use the **collections** module to do this.

In [33]:
words = ['there',
 'was',
 'a',
 'village',
 'near',
 'a',
 'jungle',
 'the',
 'village',
 'cows',
 'used',
 'to',
 'go',
 'up',
 'to',
 'the',
 'jungle',
 'in',
 'search',
 'of',
 'food.',
 'in',
 'the',
 'forest',
 'there',
 'lived',
 'a',
 'wicked',
 'lion',
 'he',
 'used',
 'to',
 'kill',
 'a',
 'cow',
 'now',
 'and',
 'then',
 'and',
 'eat',
 'her',
 'this',
 'was',
 'happening',
 'for',
 'quite',
 'sometime',
 'the',
 'cows',
 'were',
 'frightened',
 'one',
 'day',
 'all',
 'the',
 'cows',
 'held',
 'a',
 'meeting',
 'an',
 'old',
 'cow',
 'said',
 'listen',
 'everybody',
 'the',
 'lion',
 'eats',
 'one',
 'of',
 'us',
 'only',
 'because',
 'we',
 'go',
 'into',
 'the',
 'jungle',
 'separately',
 'from',
 'now',
 'on',
 'we',
 'will',
 'all',
 'be',
 'together',
 'from',
 'then',
 'on',
 'all',
 'the',
 'cows',
 'went',
 'into',
 'the',
 'jungle',
 'in',
 'a',
 'herd',
 'when',
 'they',
 'heard',
 'or',
 'saw',
 'the',
 'lion',
 'all',
 'of',
 'them',
 'unitedly',
 'moo',
 'and',
 'chased',
 'him',
 'away',
 'moral',
 'divided',
 'we',
 'fall',
 'united',
 'we',
 'stand']

b) Analyze your dicitionary by printing:
* how many keys it has 
* what the highest word frequency is
* the sum of all frequencies.

In [34]:
word_dict = {}

for word in words:
    if word.endswith("e") \
    or word.startswith("t") \
    or (word.startswith("c") and word.endswith("w")) \
    or len(word) >= 6:
        continue
    else:
        if word in word_dict:
            word_dict[word] += 1
        else:
            word_dict[word] = 1

word_dict_keys = word_dict.keys()
word_dict_values = word_dict.values()

print(f"The dictionary has {len(word_dict_keys)} keys.")
print(f"The highest word frequency is {max(word_dict_values)}.")
print(f"The sum of all frequencies is {sum(word_dict_values)}.")

The dictionary has 43 keys.
The highest word frequency is 6.
The sum of all frequencies is 69.


c) In addition, print the frequencies of the following words using your dictionary (if the word does not occur in the dictionary, print 'WORD does not occur')
* up
* near
* together
* lion
* cow

In [35]:
for word in ['up', 'near' , 'together', 'lion', 'cow']:
    if word in word_dict:
        print(f"The frequency of the word '{word}' in the dictionary is: {word_dict[word]}")
    else:
        print(f"The frequency of the word '{word}' in the dictionary is: WORD does not occur")

The frequency of the word 'up' in the dictionary is: 1
The frequency of the word 'near' in the dictionary is: 1
The frequency of the word 'together' in the dictionary is: WORD does not occur
The frequency of the word 'lion' in the dictionary is: 3
The frequency of the word 'cow' in the dictionary is: WORD does not occur


## Exercise 6: Functions

#### Exercise 6a: the beersong
Please write a function that prints the beersong when it is called.
The function:
* is called `print_beersong`
* has one positional parameter `start_number` (this is 99 in the original song) 
* prints the beer song

In [36]:
def print_beersong(start_number, /):
    """
    Prints beersong where the length of the song in determined by the given number of beer bottles.
    
    :param start_number: number of beer bottles to start the song with
    """
    for bottles in range(start_number, -1, -1): # end of range function at -1 to include 0 bottles
        remaining_num_bottles = bottles - 1
        # initialize text used for the majority of the verses
        text_bottles = "bottles"
        text_remaining_bottles = "bottles"
        key_sentence = "Take one down, pass it around," # the only sentence that changes in the last verse

        # change text accordingly to the current number of bottles
        if bottles == 2:
            text_remaining_bottles = "bottle"
        elif bottles == 1:
            text_bottles = "bottle"
            remaining_num_bottles = "no more"
        elif bottles == 0: # change text accordingly for the special final verse
            bottles = "No more"
            key_sentence = "Go to the store and buy some more," # change key sentence
            remaining_num_bottles = start_number # change to inital number of bottles

        print(f"{bottles} {text_bottles} of beer on the wall,")
        print(f"{'no more' if bottles == 'No more' else bottles} {text_bottles} of beer.") # get rid of capital letter in final verse
        print(f"{key_sentence}")
        print(f"{remaining_num_bottles} {text_remaining_bottles} of beer on the wall.")
        print()

print_beersong(99)

99 bottles of beer on the wall,
99 bottles of beer.
Take one down, pass it around,
98 bottles of beer on the wall.

98 bottles of beer on the wall,
98 bottles of beer.
Take one down, pass it around,
97 bottles of beer on the wall.

97 bottles of beer on the wall,
97 bottles of beer.
Take one down, pass it around,
96 bottles of beer on the wall.

96 bottles of beer on the wall,
96 bottles of beer.
Take one down, pass it around,
95 bottles of beer on the wall.

95 bottles of beer on the wall,
95 bottles of beer.
Take one down, pass it around,
94 bottles of beer on the wall.

94 bottles of beer on the wall,
94 bottles of beer.
Take one down, pass it around,
93 bottles of beer on the wall.

93 bottles of beer on the wall,
93 bottles of beer.
Take one down, pass it around,
92 bottles of beer on the wall.

92 bottles of beer on the wall,
92 bottles of beer.
Take one down, pass it around,
91 bottles of beer on the wall.

91 bottles of beer on the wall,
91 bottles of beer.
Take one down, pass 

#### Exercise 6b: the whatever can be in a bottle song
There are other liquids than beer than can be placed in a bottle (e.g., *99 bottles of water on the wall.*).
Please write a function that prints a variation of the beersong when it is called. All occurrences of **beer** will be replaced by what the user provides as an argument, e.g., *water*.

The function:
* is called `print_liquids`
* has one positional parameter: `start_number` (this is **99** in the original song) 
* has one keyword parameter: `liquid` (set the default value to **beer**)
* prints a liquids song

In [37]:
def print_liquids(start_number, /,liquid="beer"):
    """
    Prints song of a specified liquid where the length of the song in determined by the given number of bottles.

    :param start_number: number of bottles to start the song with
    :param liquid: liquid used for the song lyric with as default "beer"
    """
    for bottles in range(start_number, -1, -1): # end of range function at -1 to include 0 bottles
        remaining_num_bottles = bottles - 1
        # initialize text used for the majority of the verses
        text_bottles = "bottles"
        text_remaining_bottles = "bottles"
        key_sentence = "Take one down, pass it around," # the only sentence that changes in the last verse

        # change text accordingly to the current number of bottles
        if bottles == 2:
            text_remaining_bottles = "bottle"
        elif bottles == 1:
            text_bottles = "bottle"
            remaining_num_bottles = "no more"
        elif bottles == 0: # change text accordingly for the special final verse
            bottles = "No more"
            key_sentence = "Go to the store and buy some more," # change key sentence
            remaining_num_bottles = start_number # change to inital number of bottles

        print(f"{bottles} {text_bottles} of {liquid} on the wall,")
        print(f"{'no more' if bottles == 'No more' else bottles} {text_bottles} of {liquid}.") # get rid of capital letter in final verse
        print(f"{key_sentence}")
        print(f"{remaining_num_bottles} {text_remaining_bottles} of {liquid} on the wall.")
        print()

print_liquids(99, liquid="water")

99 bottles of water on the wall,
99 bottles of water.
Take one down, pass it around,
98 bottles of water on the wall.

98 bottles of water on the wall,
98 bottles of water.
Take one down, pass it around,
97 bottles of water on the wall.

97 bottles of water on the wall,
97 bottles of water.
Take one down, pass it around,
96 bottles of water on the wall.

96 bottles of water on the wall,
96 bottles of water.
Take one down, pass it around,
95 bottles of water on the wall.

95 bottles of water on the wall,
95 bottles of water.
Take one down, pass it around,
94 bottles of water on the wall.

94 bottles of water on the wall,
94 bottles of water.
Take one down, pass it around,
93 bottles of water on the wall.

93 bottles of water on the wall,
93 bottles of water.
Take one down, pass it around,
92 bottles of water on the wall.

92 bottles of water on the wall,
92 bottles of water.
Take one down, pass it around,
91 bottles of water on the wall.

91 bottles of water on the wall,
91 bottles of w

#### Exercise 6c: preprocessing text
Please write the answer to Exercise 4a as a function. The function replaces the following characters by empty strings in a text:
* newlines: '\n'
* commas: ','
* dots: '.'
* quotes: '"'

The function is:
* is called `clean_text`
* has one positional parameter `text`
* return a string, e.g., the cleaned text


In [41]:
def clean_text(text: str, /):
    """
    Preprocesses a given str text by replacing a number of punctuation characters by the empty str.
    
    :param text: str of text
    :return: str of text cleared of a set of punctuation characters
    """
    cleaned_text = text.replace('\n', '')
    cleaned_text = cleaned_text.replace(',', '')
    cleaned_text = cleaned_text.replace('.', '')
    cleaned_text = cleaned_text.replace('"', '')
    return cleaned_text

#### Exercise 6d: preprocessing text in a more general way
Please write a function that replaces all characters that the user provides by empty strings.

The function is:
* is called `clean_text_general`
* has one positional parameter `text`
* has one keyword parameter `chars_to_remove`, which is a set (set the default to {'\n', ',', '.', '"'})
* return a string, e.g., the cleaned text

When the user provides a different value to `chars_to_remove`, e.g., {'a'}, then only those characters should be replaced by empty spaces in the text.

Please store this function in a file called **assignment2_utils.py**. Please import the function and call it in this notebook.

In [42]:
from assignment2_utils import clean_text_general

In [43]:
first_alinea_of_a_story = """In a far away kingdom, there was a river. This river was home to many golden swans. The swans spent most of their time on the banks of the river. Every six months, the swans would leave a golden feather as a fee for using the lake. The soldiers of the kingdom would collect the feathers and deposit them in the royal treasury. 
One day, a homeless bird saw the river. "The water in this river seems so cool and soothing. I will make my home here," thought the bird.  """

# call the function, clearing the text from the letter "a"
check_clean_text_general = clean_text_general(first_alinea_of_a_story, {"a"})
print(check_clean_text_general)

In  fr wy kingdom, there ws  river. This river ws home to mny golden swns. The swns spent most of their time on the bnks of the river. Every six months, the swns would leve  golden fether s  fee for using the lke. The soldiers of the kingdom would collect the fethers nd deposit them in the royl tresury. 
One dy,  homeless bird sw the river. "The wter in this river seems so cool nd soothing. I will mke my home here," thought the bird.  


#### Exercise 6e: including and excluding words
Please write Exercise 5a as a function. 

The function:
* is called `exclude_and_count`
* has one positional parameter `words`, which is a list of strings.
* creates a dictionary in which the **key** is a word and the **value** is the frequency of that word. 
* words are excluded if they meet one of the following criteria:
    * end with the letter e
    * start with the letter t
    * start with the letter c and end with the letter w (both conditions must be met)
    * have six or more letters
* returns a dictionary in which the **key** is the word and the **value** is the frequency of the word. 

In [44]:
def exclude_and_count(words: list, /):
    word_dict = {}

    for word in words:
        if word.endswith("e") \
        or word.startswith("t") \
        or (word.startswith("c") and word.endswith("w")) \
        or len(word) >= 6:
            continue
        else:
            if word in word_dict:
                word_dict[word] += 1
            else:
                word_dict[word] = 1
    return word_dict

## Exercise 7: Questions about Python
Please answer the following questions about the Python language. You can give an examples if they help you.


a. What is the difference between positional parameters and keyword parameters?

In [45]:
# Positional and keyword parameters differ mostly in if you are compulsory to call them or not and 
# how you are allowed to call them.

In [46]:
# 1) Compulsory or not?
# Positional parameters are compulsory when you call a function and the values of the arguments in
# the function call are assigned to the parameters positionally. So, the first value of the argument
# in the function call is assigned to the first parameter in the function definition, the second to
# the second, third to the third, etc. 

# Keyword parameters on the otherhand are not compulsory when calling the function, as the parameters
# have a default value for each keyword parameter and are not necessarily assigned positionally.

# Let us show you by defining a function and walk you through the different parameters. 

In [47]:
# Below we have a simple function where we add numbers. The function definition has four parameters.
# There is a "/" symbol, and all parameter on the left of this symbol are positional parameters and
# all on the right are keyword parameters. As you can see, the keyword parameters have default values.

def addition(a, b, /, c=0, d=0):
    result  = a + b + c + d
    print(result)

In [48]:
# When we call the function without any arguments, we get an error stating that two positional 
# arguments are missing. So note, these positional arguments are compulsory.

addition()

TypeError: addition() missing 2 required positional arguments: 'a' and 'b'

In [49]:
# Now we call the function again. When we call the function as below, the integer 3 gets 
# assigned to a and 4 to b. This works even if we do not indicate any keyword parameters. 

addition(3, 4)

7


In [50]:
# Now we want to add four numbers and we fill in values for the keyword parameters. We
# see that the numbers add up. 

addition(3, 4, c = 1, d = 2)

10


In [51]:
# An important note is that positional parameters can also have default values:
def multiply_two_to_the_power(x=1,/):
    result = 2 ** x
    print(result)
    
multiply_two_to_the_power()

2


In [52]:
# 2) How are you allowed to call them?
# As previously mentioned, the arguments in the function call get assigned positionally 
# for the positional parameters. However, this is not the case for keyword parameters. 
# You can use the keyword in the function call and the order in how you call them does 
# not matter.

# An example:

def print_full_name(first_name, last_name, / , second_name="", third_name=""):
    print(f"My name is {first_name} {second_name} {third_name} {last_name}")
    
print_full_name("John", "Doe", third_name="Michael", second_name="Richard")

# As you can see, using the keywords, to assign a value to them, the order of the
# keyword parameters does not matter

My name is John Richard Michael Doe


In [53]:
# However, when you do not use the keywords in the function call for the 
# keyword parameters, the default is that the values of the arguments get assigned 
# positionally to the parameters, also for the keyword parameters.

print_full_name("John", "Doe", "Michael", "Richard")

My name is John Michael Richard Doe


In [54]:
# Note that you cannot use a keyword for any positional parameter, this will result
# in an error.

print_full_name(first_name="John", "Doe", "Michael", "Richard")

SyntaxError: positional argument follows keyword argument (2115595535.py, line 4)

b. What is/are the type/types of containers that you define using curly brackets?

In [55]:
# There are two types of containers that you can define using curly brackets: dictionaries and sets.
# However, when we are creating (define) an empty set or dictionary, we use different ways to do so.

In [56]:
# When creating an empty dictionary, we can just do the following two:

empty_dictionary = {} # either ...
another_empty_dictionary = dict() # or ...

print(type(empty_dictionary))
print(type(another_empty_dictionary))

<class 'dict'>
<class 'dict'>


In [57]:
# Note that even though a dictionary and set both use curly brackets, we cannot create an empty set by just using
# the '{}' as we do for a dictionay. For a set we do the following:

empty_set = set()

print(type(empty_set))

<class 'set'>


c. When comparing the containers lists and sets, there are three properties that are unique for lists. Please name and explain them.

In [58]:
# If we compare lists to sets, lists have the following three unique properties:
# 1) Lists are ordered: with this we mean that the order in which element are appended to a list determines 
# the position each element will take up in the list. To clarify, look at the following example:

list_of_words = []

list_of_words.append("first")
list_of_words.append("second")
list_of_words.append("third")

print(list_of_words)

# After printing we see that the first added element takes the first position in the list, the second takes
# takes the second and the third takes the third. So, lists are ordered and the order to which we add elements 
# to a list matters.

['first', 'second', 'third']


In [59]:
# However, when we compare this to adding elements to a set, we see something else:

set_of_words = set()

set_of_words.add("first")
set_of_words.add("second")
set_of_words.add("third")

print(set_of_words)

# When printing the set, we see that the elements are not ordered with respect to the sequence at which they
# are added. So, sets are unordered.

{'second', 'third', 'first'}


In [60]:
# 2) Lists can contain duplicates: with this we mean that when you add the same element multiple times to a list.
# Here, we show an example:

list_of_same_elements = []

list_of_same_elements.append("This will be in the list")
list_of_same_elements.append("This will be in the list")

print(list_of_same_elements)

# It shows, that we can add the exact same element multiple times to a list.

['This will be in the list', 'This will be in the list']


In [61]:
# However, for sets, this is not the case:

set_of_elements = set()

set_of_elements.add("This will be in the list")
set_of_elements.add("This will be in the list")

print(set_of_elements)

# Here we see that, even though added twice, the same element only appears once.

{'This will be in the list'}


In [62]:
# 3) The third unique property is that lists can contain, next to immutable objects, mutable objects, 
# whereas sets can only contain immutable objects. This is best shown in an example:

list_of_different_objects = []

list_of_different_objects.append("string obeject") # immutable
list_of_different_objects.append(["a", "list", "object"]) # mutable 
list_of_different_objects.append(("a", "set", "object")) # mutable 
list_of_different_objects.append(99) # immutable
list_of_different_objects.append((98,99)) # immutable

print(list_of_different_objects)

# Here we see that lists can both contain mutable as immutable elements. 

['string obeject', ['a', 'list', 'object'], ('a', 'set', 'object'), 99, (98, 99)]


In [63]:
# However, when we try to do the same for a set we run into an error:

set_of_different_objects = set()

set_of_different_objects.add("string obeject") # immutable
set_of_different_objects.add(["a", "list", "object"]) # mutable 
set_of_different_objects.add(("a", "set", "object")) # mutable 
set_of_different_objects.add(99) # immutable
set_of_different_objects.add((98,99)) # immutable

print(set_of_different_objects)

# Sets cannot contain mutable objects.

TypeError: unhashable type: 'list'

d. Explain the difference between **continue** and **break**. You can use an example for this. 

In [64]:
# Continue and break are often used in for-loops. With continue, you can skip to the next iteration once you've 
# met a particular condition in the loop. Break causes the current loop to end completely rather than skipping 
# to the next iteration. We will clarify using some examples.

In [65]:
# Assume you have a list of names and you'd like to print all those names except for the ones starting with a
# letter "J". The use of contine suits this case:

list_of_names = ["George Jr.", "George Sr.", "Sheldon", "Missy", "Meemaw", "Jeff", "Mary", "John"]

for name in list_of_names:
    if name.startswith("J"):
        continue
    else:
        print(name)
        
# As mentioned before, it can be seen that, when using continue, you move to the next iteration when a certain
# condition (in our case a name starting with the letter "J") is met. From the output it can be observed that all
# names are printed, except for those starting with the letter "J". When this condition was met, we moved to the
# next iteration.

George Jr.
George Sr.
Sheldon
Missy
Meemaw
Mary


In [66]:
# Using break, we will not move to the next iteration, but we will exit it. Assume we now have a list of animals
# and we want to print the names of these animals but stop when we found a lion.

list_of_animals = ["bear", "cat", "parrot", "dog", "sheep", "lion", "donkey", "panda"]

for animal in list_of_animals:
    if animal == "lion":
        break
    else:
        print(animal)
        
# Here we see that when we use the break statement, we exit the loop once the condition is met and we will not
# print the names of the animals that come after lion. 
# Note: depending on putting a print statement before the break or not, lion could also be printed but we just do
# not go to the next iteration.

bear
cat
parrot
dog
sheep
